### PR TITLE
Support dereferenceable attribute

### PR DESCRIPTION
**Pull request by aqjune**
This PR adds support for dereferenceable attribute at function parameters.

Dereferenceable for function's return value isn't included because it requires factoring out function attributes as a class, which will make the diff big.

LLVM unit tests: two failures added;
```
Transforms/MergeICmps/X86/entry-block-shuffled.ll
Transforms/MergeICmps/X86/multiple-blocks-does-work.ll
```

The first one does `if (x) { if (y) { ... } }` -> `if (y) { if (x) { ... } }` which is illegal because it may introduce UB.
The second one is related with memcmp's semantics; Currently, memcmp raises UB if poison bytes are compared, but this requires memcmp to return poison...

### CODE REVIEW

File: llvm_util/llvm2alive.cpp [hunk 1]
```diff
@@ -270,13 +270,30 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
...
+        attr.derefBytes = derefb;
+      }
+
       if (argI != argE) {
         // Check whether arg itr finished early because it was var arg
         if (argI->hasByValAttr())
           attr.set(ParamAttrs::ByVal);
-        else if (argI->hasReturnedAttr()) {
+
+        uint64_t derefb = argI->getType()->isPointerTy() ?
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:289]: can this not be a pointer??

File: llvm_util/llvm2alive.cpp [hunk 2]
```diff
@@ -270,13 +270,30 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
                                     !known);
     unique_ptr<Instr> ret_val;
     auto argI = fn->arg_begin(), argE = fn->arg_end();
+    auto alist = i.getAttributes();
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:273]: &

File: llvm_util/llvm2alive.cpp [hunk 3]
```diff
@@ -270,13 +270,30 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
...
+    uint64_t argidx = 0;
+
     for (auto &arg : args) {
       ParamAttrs attr;
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:277]: can we reuse handleAttribute() here?

File: ir/attrs.h [hunk 4]
```diff
@@ -28,6 +29,8 @@ class ParamAttrs final {
       ret += "readonly ";
     if (has(ParamAttrs::ReadNone))
       ret += "readnone ";
+    if (has(ParamAttrs::Dereferenceable))
```
Comment (nunoplopes) [inline ir/attrs.h:32]: this code is now conflicted, sorry

File: llvm_util/llvm2alive.cpp (hunk 1 above)
Comment (aqjune) [reply llvm_util/llvm2alive.cpp:289]: getDereferenceableBytes() fails if argI was not a pointer.

File: llvm_util/llvm2alive.cpp (hunk 3 above)
Comment (aqjune) [reply llvm_util/llvm2alive.cpp:277]: There are a few encodings that are unsupported at a callsite yet (nonnull/readnone/readonly). I think it might not be okay to show them in the printing?

Comment (aqjune) [reply llvm_util/llvm2alive.cpp:277]: I left them as a TODO comment at the place, with a minor update to the handleAttributes (so the comment makes sense)

File: ir/attrs.h (hunk 4 above)
Comment (aqjune) [reply ir/attrs.h]: Conflicts resolved

**Comment by aqjune** [comment]
The assertions revived, removing

**Comment by aqjune** [comment]
Removed. BTW, I think it is good to have invariants as explicitly stated as assertions; while implementing this I was incorrectly calling addUB() at Memory::mkInput. When I have a bandwidth I'd like to come up with a patch that adds assertions which seems to be important to me.

File: llvm_util/llvm2alive.cpp [hunk 5]
```diff
@@ -273,13 +273,32 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
...
     auto argI = fn->arg_begin(), argE = fn->arg_end();
+    const auto &alist = i.getAttributes();
+    uint64_t argidx = 0;
+
     for (auto &arg : args) {
       ParamAttrs attr;
+      // TODO: Once attributes at a call site are fully supported, we should
+      // call handleAttributes().
+
+      if (uint64_t derefb = alist.getParamDereferenceableBytes(argidx)) {
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:284]: why do you check for dereferenceable twice? once here and once below?

File: llvm_util/llvm2alive.cpp [hunk 6]
```diff
@@ -818,10 +838,9 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
     return true;
   }
 
-  optional<ParamAttrs> handleAttributes(llvm::Argument &arg) {
+  optional<ParamAttrs> handleAttributes(const llvm::AttributeSet &aset) {
     ParamAttrs attrs;
-    for (auto &attr : arg.getParent()->getAttributes()
-                         .getParamAttributes(arg.getArgNo())) {
+    for (auto &attr : aset) {
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:843]: why this change?  Sorry, I'm a bit confused with the attributes. They can come from either the call instruction or from the declaration, but I think there's an API that gives the union of those?

File: llvm_util/llvm2alive.cpp (hunk 5 above)
Comment (aqjune) [reply llvm_util/llvm2alive.cpp:284]: It is because a dereferenceable tag can be at either the callsite or at a function signature.
At a call instruction:
```
BB:
  call void @f(i32* dereferenceable(4) %p)
```
At a function signature:
```
declare void @f(i32* dereferenceable(4) %ptr)
```

File: llvm_util/llvm2alive.cpp (hunk 6 above)
Comment (aqjune) [reply llvm_util/llvm2alive.cpp:843]: The change was made to make the comment that suggests a change to use handleAttribute() (at line 281) consistent with the implementation.
For the API that returns the union of those, I also think the API should be at somewhere because it would be pretty convenient, but I couldn't find it. :/

Comment (nunoplopes) [reply llvm_util/llvm2alive.cpp:843]: see here: https://llvm.org/doxygen/Instructions_8cpp_source.html#l00321
`paramHasAttr(argno, attr`

Comment (aqjune) [reply llvm_util/llvm2alive.cpp:840]: That was helpful for removing the big if statement at visitCallInst, thanks.
However I feel I'm still not at the best answer; the main obstacle was that dereferenceable bytes should have retrieved case by case (it is possible that there exists such functionality but again I couldn't find it again :/).
Anyway, I feel if we fully support readonly/readnone/etc attached at call sites, things should become simpler. But I think it is not in high priority..

Comment (aqjune) [reply llvm_util/llvm2alive.cpp:840]: Or, to minimize diff, I can just remove this change (the change at the argument of hnadleAttributes). What do you think?

**Comment by aqjune** [comment]
Rebase (resolve conflict at transform.cpp)
